File: apps/hono/src/modules/k-char/service.ts
import { type Context } from "hono";

import * as repo from "./repo";

export async function getAll(
	c: Context,
) {
	const search =
		c.req.query("search") ?? "";

	const offset = Number(
		c.req.query("offset") ?? 0,
	);

	const limit = Number(
		c.req.query("limit") ?? 50,
	);

	const data =
		await repo.getAll(
			search,
			offset,
			limit,
		);

	return c.json(data);
}

export async function createData(
	c: Context,
) {
	const body =
		await c.req.json();

	const userId =
		c.get("userId");

	const data =
		await repo.create(
			body,
			userId,
		);

	return c.json(
		data,
		201,
	);
}

export async function editData(
	c: Context,
) {
	const id =
		c.req.param("id");

	if (!id) {
		return c.json(
			{
				message: "ID is required",
			},
			400,
		);
	}

	const body =
		await c.req.json();

	const userId =
		c.get("userId");

	const data =
		await repo.update(
			id,
			body,
			userId,
		);

	if (!data) {
		return c.json(
			{
				message:
					"Character not found",
			},
			404,
		);
	}

	return c.json(data);
}

export async function deleteData(
	c: Context,
) {
	const id =
		c.req.param("id");

	if (!id) {
		return c.json(
			{
				message: "ID is required",
			},
			400,
		);
	}

	const userId =
		c.get("userId");

	const data =
		await repo.remove(
			id,
			userId,
		);

	if (!data) {
		return c.json(
			{
				message:
					"Character not found",
			},
			404,
		);
	}

	return c.json(data);
}

export async function restoreData(
	c: Context,
) {
	const id =
		c.req.param("id");

	if (!id) {
		return c.json(
			{
				message: "ID is required",
			},
			400,
		);
	}

	const userId =
		c.get("userId");

	const data =
		await repo.restore(
			id,
			userId,
		);

	if (!data) {
		return c.json(
			{
				message:
					"Character not found",
			},
			404,
		);
	}

	return c.json(data);
}

export async function deleteDataForever(
	c: Context,
) {
	const id =
		c.req.param("id");

	if (!id) {
		return c.json(
			{
				message: "ID is required",
			},
			400,
		);
	}

	const data =
		await repo.deleteForever(id);

	return c.json(data);
}

File: apps/goapi/modules/employee/service.go
package employee

import (
	"database/sql"
	"errors"
	"net/http"
	"strconv"

	"github.com/gin-gonic/gin"
	"github.com/google/uuid"
)

type Handler struct {
	repo *Repository
}

func NewHandler(repo *Repository) *Handler {
	return &Handler{repo: repo}
}

func (h *Handler) Create(c *gin.Context) {
	var req CreateEmployeeRequest

	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	employee := &Employee{
		ID:         uuid.New(),
		EmployeeNo: req.EmployeeNo,

		FirstName:     req.FirstName,
		MiddleName:    req.MiddleName,
		LastName:      req.LastName,
		PreferredName: req.PreferredName,

		BirthDate:   req.BirthDate,
		Gender:      req.Gender,
		Nationality: req.Nationality,

		Email: req.Email,
		Phone: req.Phone,

		Address:    req.Address,
		City:       req.City,
		Province:   req.Province,
		PostalCode: req.PostalCode,
		Country:    req.Country,

		EmergencyContactName:  req.EmergencyContactName,
		EmergencyContactPhone: req.EmergencyContactPhone,

		HireDate: req.HireDate,
		Status:   EmployeeStatusActive,
	}

	if err := h.repo.Create(c, employee); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusCreated, employee)
}

func (h *Handler) GetAll(c *gin.Context) {
	limit, _ := strconv.Atoi(c.DefaultQuery("limit", "20"))
	offset, _ := strconv.Atoi(c.DefaultQuery("offset", "0"))

	employees, total, err := h.repo.GetAll(
		c,
		c.Query("search"),
		EmployeeStatus(c.Query("status")),
		limit,
		offset,
	)

	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"data": employees,
		"meta": gin.H{
			"total":  total,
			"limit":  limit,
			"offset": offset,
		},
	})
}

func (h *Handler) GetByID(c *gin.Context) {
	id, err := uuid.Parse(c.Param("id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid employee id"})
		return
	}

	employee, err := h.repo.GetByID(c, id)

	if errors.Is(err, sql.ErrNoRows) {
		c.JSON(http.StatusNotFound, gin.H{"error": "employee not found"})
		return
	}

	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, employee)
}

func (h *Handler) Update(c *gin.Context) {
	id, err := uuid.Parse(c.Param("id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid employee id"})
		return
	}

	var req UpdateEmployeeRequest

	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	employee, err := h.repo.GetByID(c, id)

	if errors.Is(err, sql.ErrNoRows) {
		c.JSON(http.StatusNotFound, gin.H{"error": "employee not found"})
		return
	}

	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	employee.FirstName = req.FirstName
	employee.MiddleName = req.MiddleName
	employee.LastName = req.LastName
	employee.PreferredName = req.PreferredName

	employee.BirthDate = req.BirthDate
	employee.Gender = req.Gender
	employee.Nationality = req.Nationality

	employee.Email = req.Email
	employee.Phone = req.Phone

	employee.Address = req.Address
	employee.City = req.City
	employee.Province = req.Province
	employee.PostalCode = req.PostalCode
	employee.Country = req.Country

	employee.EmergencyContactName = req.EmergencyContactName
	employee.EmergencyContactPhone = req.EmergencyContactPhone

	employee.HireDate = req.HireDate

	if err := h.repo.Update(c, employee); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, employee)
}

func (h *Handler) UpdateStatus(c *gin.Context) {
	id, err := uuid.Parse(c.Param("id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid employee id"})
		return
	}

	var req UpdateEmployeeStatusRequest

	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	switch req.Status {
	case EmployeeStatusActive,
		EmployeeStatusInactive,
		EmployeeStatusTerminated:
	default:
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid employee status"})
		return
	}

	if err := h.repo.UpdateStatus(c, id, req.Status); err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": err.Error()})
		return
	}

	c.Status(http.StatusNoContent)
}

func (h *Handler) Delete(c *gin.Context) {
	id, err := uuid.Parse(c.Param("id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid employee id"})
		return
	}

	if err := h.repo.Delete(c, id); err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": err.Error()})
		return
	}

	c.Status(http.StatusNoContent)
}
